added asgs-lint to its own repo, "fetch asgs-lint" now updates from there; also part of default install via asgs-brew.pl

wwlwpd opened the pull request:
To test in existing install,

Outside of ASGS:

```
git checkout issue-1426                       # this should remove bin/asgs-lint the file
./update-asgs "--run-steps asgs-lint"         # add
./update-asgs "--run-steps asgs-lint --clean" # delete
```

Inside of ASGS:
```
> fetch asgs-lint
```

To test if it actually is installed,

```
readlink -f bin/asgs-lint # will show real path if installed
```

Note:

1. There is an ever present soft link in `bin/asgs-lint` - this will probably get removed and made to be added/removed via the `fetch command. What do you think, @jasonfleming?
2. `git checkout master` after should restore `bin/asgs-lint` as it is deployed now

wwlwpd commented:
We should also try to get this right now so we can apply it to other things - `fetch` has really evolved to deal with git repositories we'd like to include (as a custom alternative to git "submodules"), but we can probably extend that to Perl modules. `build` evolved to cover things we actually compiled after fetching the source code. Idk if it makes sense to merge those, but that is just how it is partitioned in my head.

wwlwpd commented:
ok I made it add the symlink, so there is no longer a `bin/asgs-lint` that points to nothing...

jasonfleming commented:
I just tried the install process from outside the ASGS shell using the instructions from the top  of this PR, but `asgs-lint` was not installed and there was no error message. Just the normal ASGS Ready message.

wwlwpd commented:
sorry about that and thank you, I will try it from the top

wwlwpd commented:
@jasonfleming did you first checkout the branch? I reproduced the situation, but on when on a branch other than `issue-`1426`.

jasonfleming commented:
Ok I thought I did, I will double check and try again.

jasonfleming commented:
I updated to the latest on the `issue-1426` branch, and saw that it did update `bin/fetch` but I was using these commands from outside `asgsh`. Still no joy on the building `asgs-lint`:
```bash
jason@laptop-zen:~/asgs$ git branch
  issue-1244
  issue-1251
* issue-1426
  master
jason@laptop-zen:~/asgs$ git fetch --all
Fetching origin
jason@laptop-zen:~/asgs$ git pull
Already up to date.
jason@laptop-zen:~/asgs$ ./update-asgs "--run-steps asgs-lint"

you are about to run the following command:

  /home/jason/asgs/cloud/general/asgs-brew.pl --install-path=/home/jason/asgs/opt --asgs-profile=production --compiler=gfortran --machinename=debian --home=/home/jason/asgs --tmpdir=/home/jason/asgs/tmp --platform-init /home/jason/asgs/platforms/debian/init.sh --run-steps asgs-lint

proceed? [Y/n]


              █████  ███████  ██████  ███████ 
             ██   ██ ██      ██       ██      
             ███████ ███████ ██   ███ ███████ 
             ██   ██      ██ ██    ██      ██ 
             ██   ██ ███████  ██████  ███████ 

           oo__          _      _          __oo
               """--,,,_(_)_--_(_)_,,,--"""
                       _>_[_is_]_<_
               ___--""" (_)\__/(_) """--___
           oo""                            ""oo

        ██████  ███████  █████  ██████  ██    ██ ██ 
        ██   ██ ██      ██   ██ ██   ██  ██  ██  ██ 
        ██████  █████   ███████ ██   ██   ████   ██ 
        ██   ██ ██      ██   ██ ██   ██    ██       
        ██   ██ ███████ ██   ██ ██████     ██    ██ 
                             
    :.....::.....::.....::.....:::..::.....::.....:.....:
        ::::::::::::The ASGS Shell Environment:::::::::::::::
            :::::::..::::::::::::::::::::::::::::::::::::::::::::

The ASGS Shell Environment is now ready to use!

Quick Start:

   ./asgsh

Next, build ADCIRC using the command,

   build adcirc

To report bugs or request enhancements, please file them at

   https://github.com/StormSurgeLive/asgs/issues

You may also visit:

  https://tools.adcirc.live or email help@support.adcirc.live

for commercial support opportunities.

jason@laptop-zen:~/asgs$ ls -l bin/asgs-lint
ls: cannot access 'bin/asgs-lint': No such file or directory
```

wwlwpd commented:
Oh I think I see, `git branch` creates your own local branch; try this:

1. `git checkout master`
2. `git branch -d issue-1426 # deletes local branch`
3. `git fetch all            # updates branches list locally`
4. `git checkout issue-1426`

Then try the test commands.

jasonfleming commented:
Ok I tried that but got the same result as before. So I started up `asgsh` and tried `fetch asgs-lint` but got the following error message:
```
[ASGS (issue-1426)] debian@production> fetch asgs-lint
~/asgs ~/asgs
Cloning into '/home/jason/asgs/git/asgs-lint'...
git@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
Could that be the reason that the `perl-brew.pl` script did not produce the expected result? So then I got back out of `asgsh` and tried cloning the `asgs-lint` repo and it worked:
```
[ASGS (issue-1426)] debian@production> exit
exit
jason@laptop-zen:~/asgs$ git clone https://github.com/StormSurgeLive/asgs-lint
Cloning into 'asgs-lint'...
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 4 (delta 0), reused 4 (delta 0), pack-reused 0 (from 0)
Unpacking objects: 100% (4/4), 10.60 KiB | 1.77 MiB/s, done.
```

jasonfleming commented:
Will try on my desktop to see if I can reproduce.

jasonfleming commented:
Ok it turns out that the `fetch` command in `asgsh` does not work at all for me. It fails to clone the repository. I don't normally use the `git clone git@github.com:` syntax, I usually use the `https` URL (as I did above). Is there something different I need to do to enable the `git@github.com` syntax?

wwlwpd commented:
> Ok it turns out that the `fetch` command in `asgsh` does not work at all for me. It fails to clone the repository. I don't normally use the `git clone git@github.com:` syntax, I usually use the `https` URL (as I did above). Is there something different I need to do to enable the `git@github.com` syntax?

Oh, that explains it! Yes, you need to add an ssh key from the machine or environment you've installed ASGS on to your Github ssh-keys section (on the website). I strongly recommend you use ssh-keys from now on rather than https, I stopped considering https for GH interactions viable a couple of years ago, so everything I do assumes you've set up the ssh keys.

wwlwpd commented:
I wonder if there is a check or something I can do to warn people? I could check the output of `git remote -v` to look for `https` in the remote URLs, and warn?

jasonfleming commented:
Maybe just check the exit code from the `git clone`? I must have set up my GH ssh keys from all the production platforms and just forgot about that. I need to add that to the Operator onboarding checklist.

jasonfleming commented:
Ok I added my ssh keys to gh and then was able to successfully `fetch asgs-lint` within `asgsh` but when I tried a fresh install, it did not clone `asgs-lint` automatically. I also was not able to get it to work outside `asgsh`, even with the updated ssh keys on gh.

wwlwpd commented:
> Ok I added my ssh keys to gh and then was able to successfully `fetch asgs-lint` within `asgsh` but when I tried a fresh install, it did not clone `asgs-lint` automatically. I also was not able to get it to work outside `asgsh`, even with the updated ssh keys on gh.

Thank you, I will take a look so we can wrap this up!

wwlwpd commented:
testing latest change, most of is not risky but ended up being a little beyond the scope - asgs-brew clean up - but one is I deleted the "adcirc" step out of asgs-brew; we weren't using it and it'd be better with the way we're doing it now (e.g., when installing ADCIRC.Live)

wwlwpd commented:
@jasonfleming - I found the issue and fixed it. I was unable to resist the temptation to clean up `asgs-brew.pl` by removing the "adcirc" step, it was not being used. I moved the environmental variables to the ENV set up and updated the README to not mention `--build-adcirc` - we have direct commandline options to build ADCIRC outside of ASGSH via `adcirclive`.

I tested all this using a new install. It is ready for another review.

wwlwpd commented:
Quick test so you don't have to build all of ASGS,

- clone a new asgs repo
- cd into it
- `./init-asgs.sh -b -x "--run-steps asgs-lint"` 
- `./asgsh`
- `asgs-lint --help` should work:

![image](https://github.com/user-attachments/assets/a1c94371-ade3-411d-a341-cd2070b6b917)

jasonfleming commented:
Thank you @wwlwpd ! This is the right time of year for irresistible cleanups. :-) The quick test directions that you provided worked well and saved a lot of time. 

The only lingering issue I found is that when I ran `./init-asgs.sh -b -x "--run-steps asgs-lint --clean` everything was removed (as expected) except they symbolic link `bin/asgs-lint` (it was still in place and pointing to the now-deleted local `asgs-lint` git repo).

jasonfleming reviewed (approved):
Tested and confirmed that this is working as expected. Please merge. :-)
